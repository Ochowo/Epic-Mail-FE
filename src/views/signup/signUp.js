/* eslint-disable react/no-unused-prop-types */
/* eslint-disable react/forbid-prop-types */
/* eslint-disable react/prop-types */
/* eslint-disable react/destructuring-assignment */
/* eslint-disable import/no-cycle */
import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import classnames from 'classnames';
import { withRouter } from 'react-router-dom';
import { registerUser } from '../../actions/authAction';
import Logo from '../../components/logo/logo';
import Title from '../../components/title/title';
import InputFields from '../../components/input/input';
import SignIn from '../../components/signin/signIn';

class SignUp extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      login: true,
      firstName: '',
      lastName: '',
      email: '',
      password: '',
      confirmPassword: '',
      errors: {},
    };
    this.onChange = this.onChange.bind(this);
    this.loginForm = this.loginForm.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
  }

  // componentDidMount() {
  //   console.log('signup');
  //   if (this.props.isAuthenticated) {
  //     this.props.history.push('/dashboard');
  //   }
  // }

  componentWillReceiveProps(nextProps) {
    if (nextProps.errors) {
      this.setState({
        errors: nextProps.errors,
      });
    }
  }

  onChange(e) {
    const { name } = e.target;
    const { value } = e.target;
    this.setState({ [name]: value });
  }

  onSubmit(e) {
    e.preventDefault();
    const newUser = {
      firstName: this.state.firstName,
      lastName: this.state.lastName,
      email: this.state.email,
      password: this.state.password,
      confirmPassword: this.state.confirmPassword,
    };

    this.props.registerUser(newUser, this.props.history);
  }

  loginForm() {
    this.setState({ login: false });
  }

  render() {
    const { login } = this.state;
    const { errors } = this.state;
    let displayError;
    let valError;
    const err = errors.error;
    const { isAuthenticated } = this.props.auth;
    if (!isAuthenticated) {
      if (typeof err === 'string') {
        valError = errors.error;
      } else {
        displayError = errors.error;
      }
    }
    if (login) {
      return (
        <Fragment>
          <form name="epic-sign" className="box bxx" onSubmit={this.onSubmit}>
            <Logo />
            <Title
              title="
          Register on Epic Mail"
            />
            {valError && <div className="feedback">{valError}</div>}
            <InputFields
              className={classnames('input name1 name2', {
                'is-invalid': displayError,
              })}
              id="fname"
              type="text"
              placeholder="First name"
              name="firstName"
              value={this.state.firstName}
              onChange={e => this.onChange(e)}
            />
            <br />
            {displayError && <div className="feedback">{displayError.firstName}</div>}
            <InputFields
              className={classnames('input name1 name2', {
                'is-invalid': displayError,
              })}
              id="lname"
              type="text"
              placeholder="Last name"
              name="lastName"
              value={this.state.lastName}
              onChange={e => this.onChange(e)}
            />
            <br />
            {displayError && <div className="feedback">{displayError.lastName}</div>}
            <InputFields
              id="ename"
              className="input name1 email2"
              type="text"
              placeholder="Enter Email"
              name="email"
              value={this.state.email}
              onChange={e => this.onChange(e)}
            />
            <br />
            {displayError && <div className="feedback">{displayError.email}</div>}
            <InputFields
              className={classnames('input submit name1 pass2', {
                'is-invalid': displayError,
              })}
              id="pname"
              type="password"
              placeholder="Enter Password"
              name="password"
              value={this.state.password}
              onChange={e => this.onChange(e)}
            />
            <br />
            {displayError && <div className="feedback">{displayError.password}</div>}
            <InputFields
              className={classnames('input submit name1 cpass2', {
                'is-invalid': displayError,
              })}
              id="cname"
              type="password"
              placeholder="Confirm Password"
              name="confirmPassword"
              value={this.state.confirmPassword}
              onChange={e => this.onChange(e)}
            />
            <br />
            {displayError && <div className="feedback">{displayError.confirmPassword}</div>}
            <InputFields type="submit" className="logbtn" value="Sign up" />
            <button className="create name3" id="sigg" type="button" onClick={this.loginForm}>
              Login
            </button>
          </form>
        </Fragment>
      );
    }
    return <SignIn />;
  }
}
SignUp.propTypes = {
  registerUser: PropTypes.func.isRequired,
  // eslint-disable-next-line react/require-default-props
  auth: PropTypes.shape({
    root: PropTypes.string,
    errors: PropTypes.any,
    isAuthenticated: PropTypes.bool,
  }),
};
const mapStateToProps = state => ({
  auth: state.auth,
  errors: state.errors,
});
export { SignUp };
export default connect(
  mapStateToProps,
  { registerUser },
)(withRouter(SignUp));
