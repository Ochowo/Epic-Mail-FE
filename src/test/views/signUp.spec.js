import React from 'react';
import { shallow } from 'enzyme';
import { SignUp } from '../../views/signup/signUp';
import { mockStore } from '../../../mocks/mockConfig';

const store = mockStore({ auth: {} });

const props = {
  showForm: () => {},
  registerUser: () => {},
  auth: {},
  errors: {},
  store,
};
const setUp = () => {
  const component = shallow(<SignUp {...props} store={store} />);
  return component;
};
describe('SignUp component', () => {
  let component;
  beforeEach(() => {
    component = setUp();
  });
  it('should render without errors', () => {
    component.setProps({ isAuthenticated: true });
    expect(component).toMatchSnapshot();
  });

  it('ensure onchange is called for firstName', (done) => {
    const handleChangeSpy = jest.spyOn(component.instance(), 'onChange');
    const form = component.find('#fname');
    expect(form.length).toBe(1);
    const event = {
      preventDefault: jest.fn(),
      target: {
        name: 'firstName',
        value: 'test',
      },
    };
    form.simulate('change', event);
    component.instance().onChange(event);
    expect(handleChangeSpy).toHaveBeenCalled();
    done();
  });
  it('ensure onchange is called for lastName', (done) => {
    const handleChangeSpy = jest.spyOn(component.instance(), 'onChange');
    const form = component.find('#lname');
    expect(form.length).toBe(1);
    const event = {
      preventDefault: jest.fn(),
      target: {
        name: 'lastName',
        value: 'test',
      },
    };
    form.simulate('change', event);
    component.instance().onChange(event);
    expect(handleChangeSpy).toHaveBeenCalled();
    done();
  });

  it('ensure onchange is called for email', (done) => {
    const handleChangeSpy = jest.spyOn(component.instance(), 'onChange');
    const form = component.find('#ename');
    expect(form.length).toBe(1);
    const event = {
      preventDefault: jest.fn(),
      target: {
        name: 'email',
        value: 'test@gmail.com',
      },
    };
    form.simulate('change', event);
    component.instance().onChange(event);
    expect(handleChangeSpy).toHaveBeenCalled();
    done();
  });

  it('ensure onchange is called for password', (done) => {
    const handleChangeSpy = jest.spyOn(component.instance(), 'onChange');
    const form = component.find('#pname');
    expect(form.length).toBe(1);
    const event = {
      preventDefault: jest.fn(),
      target: {
        name: 'password',
        value: 'password',
      },
    };
    form.simulate('change', event);
    component.instance().onChange(event);
    expect(handleChangeSpy).toHaveBeenCalled();
    done();
  });

  it('ensure onchange is called for password', (done) => {
    const handleChangeSpy = jest.spyOn(component.instance(), 'onChange');
    const form = component.find('#cname');
    expect(form.length).toBe(1);
    const event = {
      preventDefault: jest.fn(),
      target: {
        name: 'password',
        value: 'password',
      },
    };
    form.simulate('change', event);
    component.instance().onChange(event);
    expect(handleChangeSpy).toHaveBeenCalled();
    done();
  });

  it('should call the onSubmit method', () => {
    const LoginSpy = jest.spyOn(component.instance(), 'onSubmit');
    const event = {
      preventDefault: jest.fn(),
      target: {
        name: 'test@gmail.com',
        password: 'password',
      },
    };
    component.instance().onSubmit(event);
    expect(LoginSpy).toHaveBeenCalledTimes(1);
  });
  it('should call the showForm method', () => {
    const FormSpy = jest.spyOn(component.instance(), 'loginForm');
    const form = component.find('#sigg');
    expect(form.length).toBe(1);
    form.simulate('click');
    component.instance().loginForm();
    expect(FormSpy).toBeCalled();
  });
});
