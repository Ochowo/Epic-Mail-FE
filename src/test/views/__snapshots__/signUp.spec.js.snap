// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUp component should render without errors 1`] = `ShallowWrapper {}`;
